tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+#  (PART) Conclusion {-}
+
+# Tradeoffs
+
+::: {.alert .alert-success} 
+This somewhat opinionated discussion was added by [Max Held](https://www.maxheld.de).
+Your mileage may vary.
+:::
+
+Test doubles such as mocks or fakes can help you isolate your tests from the state of external HTTP API.
+This can be a good thing, because it allows you to test your code against a rare responses, or when the API is not available.
+
+Unfortunately, mocking and faking add their own complexity to a project.
+Respectively, you have to carry expected responses or even (rudimentary) API behavior in your code base.
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
```

Comment:
It might make sense to be less sure e.g. "it might help with not having to test all functions with the API" or so.

Reply:
Isn't "can help" already careful?
I like simple, opinionated heuristics.